crates/cli/src/native_query/type_solver/mod.rs

```diff
@@ -24,73 +22,72 @@ pub fn unify(
     configuration: &Configuration,
     required_type_variables: &[TypeVariable],
     object_type_constraints: &mut BTreeMap<ObjectTypeName, ObjectTypeConstraint>,
-    mut type_variables: HashMap<TypeVariable, HashSet<TypeConstraint>>,
+    type_variables: HashMap<TypeVariable, BTreeSet<TypeConstraint>>,
 ) -> Result<(
     HashMap<TypeVariable, Type>,
     BTreeMap<ObjectTypeName, ObjectType>,
 )> {
     let mut added_object_types = BTreeMap::new();
     let mut solutions = HashMap::new();
+    let mut substitutions = HashMap::new();
     fn is_solved(solutions: &HashMap<TypeVariable, Type>, variable: TypeVariable) -> bool {
         solutions.contains_key(&variable)
     }
 
     #[cfg(test)]
     println!("begin unify:\n  type_variables: {type_variables:?}\n  object_type_constraints: {object_type_constraints:?}\n");
 
-    // TODO: This could be simplified. Instead of mutating constraints using `simplify_constraints`
-    // we might be able to roll all constraints into one and pass that to `constraint_to_type` in
-    // one step, but leave the original constraints unchanged if any part of that fails. That could
-    // make it simpler to keep track of source locations for when we want to report type mismatch
-    // errors between constraints.
     loop {
         let prev_type_variables = type_variables.clone();
         let prev_solutions = solutions.clone();
+        let prev_substitutions = substitutions.clone();
```

Comment:
I'm glad your simplification was instead to make a clone of the substitutions for backtracking.